programs/clearing_house/src/math/margin.rs

```diff
@@ -1,91 +1,260 @@
 // use crate::math::repeg;
 use crate::math::slippage::calculate_slippage;
+use crate::state::bank::Bank;
 use crate::state::bank::BankBalanceType;
 use crate::state::bank_map::BankMap;
+use crate::state::market::Market;
 use crate::state::market_map::MarketMap;
+use crate::state::oracle::OraclePriceData;
 use crate::state::oracle_map::OracleMap;
 use crate::state::state::OracleGuardRails;
+use crate::state::user::{MarketPosition, UserBankBalance};
 use crate::validate;
+use num_integer::Roots;
 use solana_program::msg;
+use std::cmp::{max, min};
 use std::ops::Div;
 
-#[derive(Copy, Clone)]
+#[derive(Clone, Copy, PartialEq, Debug)]
 pub enum MarginRequirementType {
     Initial,
     Partial,
     Maintenance,
 }
 
+pub fn calculate_size_markup_liability_weight(
+    size: u128, // AMM_RESERVE_PRECISION
+    imf_factor: u128,
+    liability_weight: u128,
+    precision: u128,
+) -> ClearingHouseResult<u128> {
+    if imf_factor == 0 {
+        return Ok(liability_weight);
+    }
+
+    let size_sqrt = ((size / 1000) + 1).nth_root(2); //1e13 -> 1e10 -> 1e5
+
+    let liability_weight_numer = liability_weight
+        .checked_sub(
+            liability_weight
+                .checked_div(max(1, BANK_IMF_PRECISION / imf_factor))
+                .ok_or_else(math_error!())?,
+        )
+        .ok_or_else(math_error!())?;
+
+    // increases
+    let size_markup_liability_weight = liability_weight_numer
+        .checked_add(
+            size_sqrt // 1e5
+                .checked_mul(imf_factor)
+                .ok_or_else(math_error!())?
+                .checked_div(100_000 * BANK_IMF_PRECISION / precision) // 1e5 * 1e2
+                .ok_or_else(math_error!())?,
+        )
+        .ok_or_else(math_error!())?;
+
+    let max_liability_weight = max(liability_weight, size_markup_liability_weight);
+    Ok(max_liability_weight)
+}
+
+pub fn calculate_size_discount_asset_weight(
+    size: u128, // AMM_RESERVE_PRECISION
+    imf_factor: u128,
+    asset_weight: u128,
+) -> ClearingHouseResult<u128> {
+    if imf_factor == 0 {
+        return Ok(asset_weight);
+    }
+
+    let size_sqrt = (size / 1000).nth_root(2) + 1; //1e13 -> 1e10 -> 1e5
+    let imf_numerator = BANK_IMF_PRECISION + BANK_IMF_PRECISION / 10;
+
+    let size_discount_asset_weight = imf_numerator
+        .checked_mul(BANK_WEIGHT_PRECISION)
+        .ok_or_else(math_error!())?
+        .checked_div(
+            BANK_IMF_PRECISION
+                .checked_add(
+                    size_sqrt // 1e5
+                        .checked_mul(imf_factor)
+                        .ok_or_else(math_error!())?
+                        .checked_div(100_000) // 1e5
+                        .ok_or_else(math_error!())?,
+                )
+                .ok_or_else(math_error!())?,
+        )
+        .ok_or_else(math_error!())?;
+
+    let min_asset_weight = min(asset_weight, size_discount_asset_weight);
+
+    Ok(min_asset_weight)
+}
+
+pub fn calculate_bank_equity_value(
```

Comment:
feel like is it weird to call this `equity` value since it can also be a liability when it's a borrow. (equity is assets minus liabilities). Could call it `calculate_bank_balance_value` ?